Persistent Stats: persist stats history to disk

miasantreble opened the pull request:
This PR continues the work in https://github.com/facebook/rocksdb/pull/4748 and https://github.com/facebook/rocksdb/pull/4535 by adding a new DBOption `persist_stats_to_disk` which instructs RocksDB to persist stats history to RocksDB itself. When statistics is enabled, and  both options `stats_persist_period_sec` and `persist_stats_to_disk` are set, RocksDB will periodically write stats to a built-in column family in the following form: key -> (timestamp in microseconds)#(stats name), value -> stats value. The existing API `GetStatsHistory` will detect the current value of `persist_stats_to_disk` and either read from in-memory data structure or from the hidden column family on disk.

riversand963 commented on db/db_impl_open.cc:
```diff
@@ -409,6 +420,15 @@ Status DBImpl::Recover(
+    // When recovering from a DB which already contains persistent stats CF,
+    // the CF is already created in VersionSet::ApplyOneVersionEdit, but
+    // column family handle was not. Need to explicitly create handle here.
+    if (mutable_db_options_.persist_stats_to_disk &&
```
Is this a duplicate with line 390?

miasantreble replied:
Yes, good catch

riversand963 commented on db/version_set.cc:
```diff
@@ -3386,11 +3386,22 @@ Status VersionSet::ApplyOneVersionEdit(
+      // recover persistent_stats CF from a DB that already contains it
+      // TODO(Zhongyi): find out a way to pass cf options
+      if (is_persistent_stats_column_family) {
+        cfd = CreateColumnFamily(ColumnFamilyOptions(), &edit);
```
Are we going to use default values for the persistent stats column family? I feel many option values for persistent stats column family can differ from the default because the column family itself is kind of special.

miasantreble commented on db/db_impl.cc:
```diff
@@ -644,7 +653,7 @@ void DBImpl::StartTimedTasks() {
 
 // esitmate the total size of stats_history_
-size_t DBImpl::EstiamteStatsHistorySize() const {
+size_t DBImpl::EstimateStatsHistorySize() const {
```
this is the in-memory stats history which is planned to coexist with on-disk stats history. Will be renamed in the next commit

miasantreble commented on db/db_impl_open.cc:
```diff
@@ -383,11 +383,22 @@ Status DBImpl::Recover(
   }
   if (s.ok() && !read_only) {
     for (auto cfd : *versions_->GetColumnFamilySet()) {
+      // skip directory creation if persistent_stats CF was just created
```
because it's already created in https://github.com/facebook/rocksdb/blob/master/db/db_impl.cc#L1972 as part of the call to `CreatePersistStatsHandle()->CreateColumnFamily()`

miasantreble replied:
there are two code paths when it comes to creating/recovering persistent stats column family:
* creating for the first time: `InitPersistStatsColumnFamily` calls `DBImpl::CreateColumnFamilyImpl`  which add directories (see https://fburl.com/azkputvf) for newly created column family
* recovering: `InitPersistStatsColumnFamily` just creates cf handle and no directories is added
We need to skip re-creating directories for the first case which is why we handle it differently

siying replied:
It seems to be wrong that we first do 2 and then 1. If we can recover stats CF, we should recover from there. Only if itr is not there, we should create a new one.

miasantreble replied:
Makes sense, will move the call `InitPersistStatsColumnFamily` to after creating directories.

siying commented on db/db_impl_open.cc:
```diff
@@ -518,6 +531,37 @@ Status DBImpl::Recover(
   return s;
 }
 
+Status DBImpl::CreatePersistStatsHandle() {
```
This function name is misleading to me. In some cases, it's not just to create the handle but the column family too.

siying commented on include/rocksdb/options.h:
```diff
@@ -683,6 +683,12 @@ struct DBOptions {
   // Default: 600
   unsigned int stats_persist_period_sec = 600;
 
+  // If true, automatically persist stats to a hidden column family every
```
Need to mention the name of the hidden column family name here. Also need to mention in case a user provide a name of the name, what the behavior they are expected to see.

miasantreble replied:
In `DBImpl::Recover` where we recover the manifest, how do we tell if the hidden column family was explicitly created by user or implicitly created by persistent stats feature? 
I imagine if we want to disable persistent stats when user provided column family name conflicts with hidden column family name, we need to be able to tell the difference?

siying commented on db/db_impl.cc:
```diff
@@ -101,6 +102,7 @@
 
 namespace rocksdb {
 const std::string kDefaultColumnFamilyName("default");
+const std::string kPersistentStatsColumnFamilyName("_stats_history");
```
Can we make it even less likely to cause a conflict? For example, put "rocksdb" as a part of the name, and having more "_" than just a leading one?

siying commented on db/db_options_test.cc:
```diff
@@ -797,6 +798,210 @@ TEST_F(DBOptionsTest, InMemoryStatsHistoryPurging) {
+    // for (const auto& kv : stats_map) {
+    //   fprintf(stdout, "key = %s, value = %s\n", kv.first.c_str(),
+    //   ToString(kv.second).c_str());
+    // }
```
Remove.

siying commented on db/db_options_test.cc:
```diff
@@ -797,6 +798,210 @@ TEST_F(DBOptionsTest, InMemoryStatsHistoryPurging) {
+    // if (count % 100 == 0) {
+    //   fprintf(stdout, "key = %s, value = %s\n",
+    //   iter->key().ToString().c_str(), iter->value().ToString().c_str());
+    // }
```
Remove the commented lines.

siying replied:
It's good that you do some manual testing and make sure the outputs are correct. It's a good idea to at least put some of it as a part of the test itself. We need to make sure the data itself in the stat CF is correct, as part of the regression test.

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,102 @@
+#include "util/string_util.h"
+
+namespace rocksdb {
+const int kTimestampStringMinimumLength = 16;
```
Should be maximum?

siying commented on db/db_impl.cc:
```diff
@@ -674,17 +683,34 @@ void DBImpl::PersistStats() {
+    now_micros_string.insert(
+        0, kTimestampStringMinimumLength - now_micros_string.length(), '0');
+    int keycount = 0;
+    WriteOptions wo;
```
Do we consider to put w.low_pri=true and w.no_slowdown to be true? It's also a good idea to explicitly put wo.sync to be false. We may even consider whether we want to write to WAL.

siying commented on db/db_impl.cc:
```diff
@@ -674,17 +683,34 @@ void DBImpl::PersistStats() {
+    // for (auto iter = stats_map.begin(); iter != stats_map.end(); ++iter) {
+    for (const auto& stat : stats_map) {
+      std::string key(now_micros_string);
+      key.append("#").append(stat.first);
```
In this code base, we tend to use sprintf to do this kind of formatting. In this place, I also feel that it is clearer to read.

miasantreble replied:
the issue here is that key's length is unknown so we need to guess the length when defining the char array, which might waste some memory, and that's for every stat that we collect.

siying replied:
It's just waste temporary memory here, not every stat. Do I understand correctly?

miasantreble replied:
yes, just temporary memory waste, will replace std::string with char array and use snprintf

siying commented on db/db_impl.cc:
```diff
@@ -679,12 +689,45 @@ void DBImpl::PersistStats() {
+
+  if (immutable_db_options_.persist_stats_to_disk) {
+    char* timestamp = new char[kNowMicrosStringLength + 1];
+    memset(timestamp, 0, kNowMicrosStringLength + 1);
```
Do we need to do it at all?

siying commented on db/db_impl.cc:
```diff
@@ -679,12 +689,45 @@ void DBImpl::PersistStats() {
+  }
+
+  if (immutable_db_options_.persist_stats_to_disk) {
+    char* timestamp = new char[kNowMicrosStringLength + 1];
```
Why not `char timestamp[kNowMicrosStringLength + 1];`?

miasantreble replied:
visual studio compiler doesn't like it, see https://ci.appveyor.com/project/Facebook/rocksdb/builds/24222890 for commit https://github.com/facebook/rocksdb/pull/5046/commits/a00f70e89f9af4107d38615ee95e4832a21473af

siying replied:
Search the code base, you'll see that it is used in a lot of places. For example:
https://github.com/facebook/rocksdb/blob/05d9d82181547703e5610f423be7da6bfda3ac00/cache/sharded_cache.cc#L134-L139

miasantreble replied:
I think it's because MSVS got confused and thought `kNowMicrosStringLength` is not a constant. It might have to do with it being an extern and defined elsewhere

siying replied:
Why not make it a const?

miasantreble replied:
it is a const, the problem is MSVS thinks otherwise

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,105 @@
+    auto iter =
+        db_impl_->NewIterator(ro, db_impl_->PersistentStatsColumnFamily());
+    char timestamp[kNowMicrosStringLength + 1];
+    memset(timestamp, 0, kNowMicrosStringLength + 1);
```
Same here. Do we need it?

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,105 @@
+  // try to find next entry in stats_history_ map
+  if (db_impl_ != nullptr) {
+    ReadOptions ro;
+    auto iter =
```
Spell out the type. This `auto` doesn't help readability. At least do `auto*`.

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,105 @@
+      delete iter;
+      return;
+    }
+    time_ = parseKey(iter->key().ToString(), start_time).first;
```
ToString() does a memory copy. If it is not so hard, we should avoid it.

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,105 @@
+    // find all entries with timestamp equal to time_
+    std::map<std::string, uint64_t> new_stats_map;
+    std::pair<uint64_t, std::string> parsed_key =
+        parseKey(iter->key().ToString(), start_time);
```
Same here.

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,105 @@
+  return stats_map_;
+}
+
+std::pair<uint64_t, std::string> parseKey(std::string key,
```
At least `const std::string&`. Ideally, we should use a Slice.

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,105 @@
+    result.second = "";
+  } else {
+    uint64_t parsed_time =
+        static_cast<uint64_t>(std::stoull(key.substr(0, pos)));
```
Is there a reason we don't use `ConsumeDecimalNumber()` or `ParseUint64()`?

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,105 @@
+    } else {
+      result.first = parsed_time;
+      std::string key_resize = key.substr(pos + 1);
+      key_resize.erase(std::find(key_resize.begin(), key_resize.end(), '\0'), key_resize.end());
```
What is it for?

miasantreble replied:
it's a fail-safe to handle the case that key was improperly constructed with multiple '\0' at the end or there are garbage after the first '\0'.

miasantreble replied:
this line will be removed

siying commented on db/db_impl.cc:
```diff
@@ -679,12 +689,45 @@ void DBImpl::PersistStats() {
+    if (stats_slice_initialized_) {
+      for (const auto& stat : stats_map) {
+        char key[100];
+        memset(key, 0, 100);
```
Do we need it?

siying commented on db/db_impl.cc:
```diff
@@ -679,12 +689,45 @@ void DBImpl::PersistStats() {
+        if (stats_slice_.find(stat.first) != stats_slice_.end()) {
+          uint64_t delta = stat.second - stats_slice_[stat.first];
+          // TODO(Zhongyi): add counters for failed writes
+          batch.Put(persist_stats_cf_handle_, Slice(key, std::min(100, length)),
```
Need to document in options.h that stat name is truncated to 100 bytes.

siying commented on db/db_impl.cc:
```diff
@@ -664,6 +673,7 @@ size_t DBImpl::EstiamteStatsHistorySize() const {
 
 void DBImpl::PersistStats() {
   TEST_SYNC_POINT("DBImpl::PersistStats:Entry");
+  // 16 digit microseconds timestamp => [Sep 9, 2001 ~ Nov 20, 2286]
```
Do we really need microsecond level timestamp in stats?

miasantreble replied:
so shall we use millisecond or second for timestamps?
I guess second level is sufficient?

siying commented on db/db_impl.cc:
```diff
@@ -936,6 +985,7 @@ Status DBImpl::SetDBOptions(
           mutex_.Lock();
         }
         if (new_options.stats_persist_period_sec > 0) {
+          InitPersistStatsColumnFamily();
```
Why do we need to call it again here while we already called it in DBImpl::Open()?

miasantreble replied:
this was added earlier when persist_to_disk was a mutable cf option, so we need to handle the case of after upgrading, user dynamically enabled on-disk persistent stats, which is not the case anymore, will be deleted

siying commented on db/version_set.cc:
```diff
@@ -3916,11 +3916,22 @@ Status VersionSet::ApplyOneVersionEditToBuilder(
+      // recover persistent_stats CF from a DB that already contains it
+      // TODO(Zhongyi): find out a way to pass cf options
+      if (is_persistent_stats_column_family) {
+        cfd = CreateColumnFamily(ColumnFamilyOptions(), &edit);
```
Here is a hidden contract here. We need to guarantee that we are able to open all the future persistant stats CF, and the old one. I think we need to think through what we should do to guarantee this kind of compatibility. And we need to document it in the code. Maybe it's a good idea to reserve some versioning of the persistent stats CF, so that we reserve the possibility of doing format migration in the future.

miasantreble replied:
version key added, currently will accept anything with a version value >= 0, which we will reject user-defined column family with the same name (unless it also have a version key and value which is unlikely); in the future if we ever change persistent stats key/value encoding, we will be able to read the version number and use different parsing code  in `parseKey`

miasantreble commented on db/version_set.cc:
```diff
@@ -3916,11 +3916,22 @@ Status VersionSet::ApplyOneVersionEditToBuilder(
+      // recover persistent_stats CF from a DB that already contains it
+      // TODO(Zhongyi): find out a way to pass cf options
+      if (is_persistent_stats_column_family) {
+        cfd = CreateColumnFamily(ColumnFamilyOptions(), &edit);
```
Plan to add `ColumnFamilyOptions::OptimizeForPersistentStats()` to create a customized cfoption

miasantreble commented on db/db_impl.cc:
```diff
@@ -679,12 +688,45 @@ void DBImpl::PersistStats() {
     }
+    stats_slice_initialized_ = true;
+    std::swap(stats_slice_, stats_map);
+    Write(rocksdb::WriteOptions(), &batch);
```
will reuse WriteOptions from above

miasantreble commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,106 @@
+      result.first = parsed_time;
+      std::string key_resize = key_str.substr(pos + 1);
+      key_resize.erase(std::find(key_resize.begin(), key_resize.end(), '\0'), key_resize.end());
+      result.second = key_resize;
```
I can add a static function in `PersistentStatsHistoryIterator` and call it from `DBImpl::PersistStats()`

miasantreble commented:
> Do we plan to find the way to prevent stats CF from holding older log files from being deleted?

will be done in a follow up PR to always update log_number for persistent stats column family even if there is no data from it in the memtable

siying commented on db/db_impl/db_impl.cc:
```diff
@@ -685,12 +694,49 @@ void DBImpl::PersistStats() {
+      // write version key
+      char version_key[100];
+      int version_key_length = PersistentStatsHistoryIterator::EncodeVersionKey(
+          now_micros, 100, version_key);
```
You are doing a version key per stats. Do you consider to have one single version key for the whole CF?

siying commented on db/persistent_stats_history.h:
```diff
@@ -0,0 +1,80 @@
+  std::map<std::string, uint64_t> stats_map_;
+  Status status_;
+  bool valid_;
+  int version_;
```
I don't see where the version key is read out and check whether the version is supported. The purpose of the format version is to identify unsupported version so that it can skip.

miasantreble replied:
currently it's in `DecodeVersionNumber` and called indirectly by `PersistentStatsHistoryIterator::Next()` is called. If the version number is not found, the iterator will be invalidated

siying replied:
That still doesn't make it stop reading from a future format, which may cause problems.

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,156 @@
+
+int PersistentStatsHistoryIterator::EncodeKey(uint64_t now_micros,
+                                              const std::string& key, int size,
+                                              char* buf) {
```
I don't get why encoder belongs to iterator. Iterator is only for read, right?

miasantreble replied:
Yes but you said earlier 
> I prefer the key encoding code is placed closer to key decoding code. It makes easier for people to read the code and reason about correctness.

will think about a better place to put them

siying commented on db/persistent_stats_history.cc:
```diff
@@ -0,0 +1,156 @@
+      result.first = parsed_time;
+      std::string key_resize = key_str.substr(pos + 1);
+      auto it = std::find(key_resize.begin(), key_resize.end(), '\0');
+      assert(it == key_resize.end());
```
I'm thinking that assertion may be too strong for this. It is true that RocksDB is supposed to write to this CF and now RocksDB reads from it. But writing to it might happen by another version and may even from another host. Asserting data correctness of a DB not generated by the current process sounds not right to me.

miasantreble replied:
sure, will remove the assert and see if it's possible to move the check in a unit test

siying commented on include/rocksdb/options.h:
```diff
@@ -92,6 +92,8 @@ struct ColumnFamilyOptions : public AdvancedColumnFamilyOptions {
   ColumnFamilyOptions* OptimizeForSmallDb(
       std::shared_ptr<Cache>* cache = nullptr);
 
+  ColumnFamilyOptions* OptimizeForPersistentStats();
```
Do we have to make it a public function?

siying commented on include/rocksdb/options.h:
```diff
@@ -694,6 +696,16 @@ struct DBOptions {
+  // stats_persist_period_sec seconds; otherwise, write to an in-memory
+  // struct. User can query through `GetStatsHistory` API.
+  // If user attempt to create a column family on a DB which have previously
+  // set persist_stats_to_disk to true, the column family creation will fail.
```
Do you mean creating a column family with the same name?

siying commented on options/options.cc:
```diff
@@ -506,7 +506,17 @@ ColumnFamilyOptions* ColumnFamilyOptions::OptimizeForSmallDb(
+  target_file_size_base = 2 * 1048576;
+  max_bytes_for_level_base = 10 * 1048576;
+  snap_refresh_nanos = 0;
+  soft_pending_compaction_bytes_limit = 256 * 1048576;
```
This feels too small to me.

siying commented on options/options.cc:
```diff
@@ -506,7 +506,17 @@ ColumnFamilyOptions* ColumnFamilyOptions::OptimizeForSmallDb(
+  max_bytes_for_level_base = 10 * 1048576;
+  snap_refresh_nanos = 0;
+  soft_pending_compaction_bytes_limit = 256 * 1048576;
+  hard_pending_compaction_bytes_limit = 1073741824ul;
```
This also feels too small.

siying commented on db/db_impl/db_impl_open.cc:
```diff
@@ -529,6 +536,40 @@ Status DBImpl::Recover(
+  if (!persistent_stats_cfd_exists) {
+    mutex_.Unlock();
+    ColumnFamilyHandle* handle = nullptr;
+    Status s = CreateColumnFamily(ColumnFamilyOptions(),
```
Here still default `ColumnFamilyOptions()` is used without calling `OptimizeForPersistentStats()`.

siying commented on db/persistent_stats_history.h:
```diff
@@ -0,0 +1,80 @@
+  // Encode timestamp and version key into buf
+  // Format: timestamp(10 digit) + '#' + kVersionKeyString
+  // Total length of encoded key will be capped at 100 bytes
+  static int EncodeVersionKey(uint64_t timestamp, int size, char* buf);
```
I suggest we call it "format version", rather than "version" to avoid ambiguity.

miasantreble commented:
the latest commit d4ba50d addressed most issues except upgrade/downgrade compatibility check. The format version is now a single key instead of timestamp+'#'+key, and the check is moved to DB::Open code path so unsupported data will not be read at all

siying commented on include/rocksdb/options.h:
```diff
@@ -694,6 +696,17 @@ struct DBOptions {
+  // struct. User can query through `GetStatsHistory` API.
+  // If user attempt to create a column family with the same name on a DB which
+  // have previously set persist_stats_to_disk to true, the column family
+  // creation will fail.
```
And what will happen to the persistent stats?

siying commented on include/rocksdb/stats_history.h:
```diff
@@ -53,6 +53,8 @@ class StatsHistoryIterator {
   // REQUIRES: Valid()
   virtual uint64_t GetStatsTime() const = 0;
 
+  virtual int GetFormatVersion() const { return -1; }
```
Why do we need to expose it to users at all?

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,176 @@
+    return -1;
+  }
+  char key_string[100];
+  GetVersionKeyString(key_string, type);
```
Similarly, it's too complicated. If we make `kFormatVersionKeyString` an std::string, maybe we just can do `Get(options, kFormatVersionKeyString)`.

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,176 @@
+  std::string result;
+  Status s = db->Get(options, db->PersistentStatsColumnFamily(),
+                     Slice(key_string), &result);
+  if (s.IsNotFound() || !s.ok()) {
```
!s.ok() includes IsNotFound().

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,176 @@
+
+  // read version_number but do nothing in current version
+  int version_number =
+      static_cast<int>(std::stoull(result));
```
Can we use `ParseUint64()`?

miasantreble commented on db/db_impl/db_impl.cc:
```diff
@@ -101,6 +102,8 @@
 namespace rocksdb {
 const std::string kDefaultColumnFamilyName("default");
+const std::string kPersistentStatsColumnFamilyName(
+    "___rocksdb_reserved_stats_history");
```
removed

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,171 @@
+const std::string kFormatVersionKeyString =
+    "__persistent_stats_format_version__";
+const std::string kReaderVersionKeyString =
+    "__persistent_stats_reader_version__";
```
Can we rename it from "reader version" to something like "compatible version" or something like that?

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,171 @@
+  if (type >= StatsVersionKeyType::kKeyTypeMax) {
+    return -1;
+  }
+  std::string key = GetVersionKeyString(type);
```
Can we directly pass in the key name, instead of calling such a function?

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  // read version_number but do nothing in current version
+  int version_number = static_cast<int>(ParseUint64(result));
```
There is a `ParseInt()` function near `ParseUint64()`......

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,171 @@
+  options.verify_checksums = true;
+  std::string result;
+  Status s = db->Get(options, db->PersistentStatsColumnFamily(), key, &result);
+  if (!s.ok() || result.length() == 0) {
```
Do we really want to swallow the error? At least we should log in.

miasantreble replied:
changed return type to Status and have the caller log the errors

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,171 @@
+      if (UNLIKELY(kv.first != time_)) {
+        break;
+      }
+      if (UNLIKELY(kv.second.compare(kFormatVersionKeyString) == 0)) {
```
I'm not big fan of UNLIKELY() hint. Unless it's in a critical path and we really need this optimization, I am against using it.

siying commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,171 @@
+      if (UNLIKELY(kv.second.compare(kFormatVersionKeyString) == 0)) {
+        continue;
+      }
+      new_stats_map[kv.second] = std::stoull(iter->value().ToString());
```
Can you use `ParseUint64()`?

siying commented on include/rocksdb/options.h:
```diff
@@ -694,6 +694,18 @@ struct DBOptions {
+  // family name: ___rocksdb_stats_history___) every
+  // stats_persist_period_sec seconds; otherwise, write to an in-memory
+  // struct. User can query through `GetStatsHistory` API.
+  // If user attempt to create a column family with the same name on a DB which
```
attempts

riversand963 commented on db/db_impl/db_impl.cc:
```diff
@@ -485,10 +489,15 @@ Status DBImpl::CloseHelper() {
     mutex_.Unlock();
-    delete default_cf_handle_;
+    if (default_cf_handle_) {
+      delete default_cf_handle_;
```
Set both `default_cf_handle_` and `persist_stats_cf_handle_` to `nullptr` after deletion?

riversand963 commented on db/db_impl/db_impl.cc:
```diff
@@ -744,8 +782,13 @@ Status DBImpl::GetStatsHistory(
   }
-  stats_iterator->reset(
-      new InMemoryStatsHistoryIterator(start_time, end_time, this));
+  if (immutable_db_options_.persist_stats_to_disk) {
```
Probably not necessarily in this PR. In the future, it would be nice if we can have `StatsHistoryIterator` unifying the in-memory, on-disk case. Or we can get rid of the in-memory case.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -499,6 +507,97 @@ Status DBImpl::Recover(
+    // We are recovering from a DB which already contains persistent stats CF,
+    // the CF is already created in VersionSet::ApplyOneVersionEdit, but
+    // column family handle was not. Need to explicitly create handle here.
+    persist_stats_cf_handle_ = new ColumnFamilyHandleImpl(
```
We can avoid calling `versions_->GetColumnFamilySet()->GetColumnFamily(kPersistentStatsColumnFamilyName)` twice.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -499,6 +507,97 @@ Status DBImpl::Recover(
+      should_persist_version = true;
+    }
+  }
+  if (should_persist_version) {
```
Check the value of `s` too. Like
```
if (s.ok() && should_persist_format_version)
```

riversand963 commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,172 @@
+  options.verify_checksums = true;
+  std::string result;
+  Status s = db->Get(options, db->PersistentStatsColumnFamily(), key, &result);
+  if (!s.ok() || result.length() == 0) {
```
result.empty()

riversand963 commented on monitoring/persistent_stats_history.cc:
```diff
@@ -0,0 +1,172 @@
+  // TODO(Zhongyi): add counters to track parse failures?
+  if (pos == std::string::npos) {
+    result.first = port::kMaxUint64;
+    result.second = "";
```
Nit: result.second.clear()

riversand963 commented on db/db_impl/db_impl.cc:
```diff
@@ -634,22 +646,24 @@ void DBImpl::StartTimedTasks() {
         thread_dump_stats_.reset(new rocksdb::RepeatableThread(
             [this]() { DBImpl::DumpStats(); }, "dump_st", env_,
-            stats_dump_period_sec * 1000000));
+            static_cast<uint64_t>(stats_dump_period_sec) *
```
I think only one `static_cast` suffices.

riversand963 commented on db/db_impl/db_impl.cc:
```diff
@@ -634,22 +646,24 @@ void DBImpl::StartTimedTasks() {
         thread_persist_stats_.reset(new rocksdb::RepeatableThread(
             [this]() { DBImpl::PersistStats(); }, "pst_st", env_,
-            stats_persist_period_sec * 1000000));
+            static_cast<uint64_t>(stats_persist_period_sec) *
```
Ditto.

riversand963 commented on db/db_impl/db_impl.cc:
```diff
@@ -946,7 +995,8 @@ Status DBImpl::SetDBOptions(
           thread_dump_stats_.reset(new rocksdb::RepeatableThread(
               [this]() { DBImpl::DumpStats(); }, "dump_st", env_,
-              new_options.stats_dump_period_sec * 1000000));
+              static_cast<uint64_t>(new_options.stats_dump_period_sec) *
```
One `static_cast` should be sufficient.

riversand963 commented on db/db_impl/db_impl.cc:
```diff
@@ -961,7 +1011,8 @@ Status DBImpl::SetDBOptions(
           thread_persist_stats_.reset(new rocksdb::RepeatableThread(
               [this]() { DBImpl::PersistStats(); }, "pst_st", env_,
-              new_options.stats_persist_period_sec * 1000000));
+              static_cast<uint64_t>(new_options.stats_persist_period_sec) *
```
Ditto.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -496,6 +504,98 @@ Status DBImpl::Recover(
+  Status s;
+  mutex_.Unlock();
+  // persist version when stats CF doesn't exist
+  bool should_persist_format_version = !persistent_stats_cfd_exists_;
```
Accessing `persistent_stats_cfd_exists_` after releasing db mutex?

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1176,6 +1180,19 @@ class DBImpl : public DB {
+  // for compatibility check. Write format version if it's created for the
+  // first time, read format version and check compatibility if recovering
+  // from disk
+  Status PersistentStatsProcessFormatVersion();
```
Optional : mention that this method requires DB mutex at its entrance, but may release and re-acquire the db mutex in its execution.
